farm/data_handler/processor.py

```diff
@@ -2983,173 +2934,226 @@ def file_to_dicts(self, file: str) -> [dict]:
+                    if self.shuffle_negatives:
+                        random.shuffle(hard_negative_context)
+                    hard_negative_context = hard_negative_context[:self.num_hard_negatives]
+
+                    positive_ctx_titles = [passage.get("title", None) for passage in positive_context]
+                    positive_ctx_texts = [passage["text"] for passage in positive_context]
+                    hard_negative_ctx_titles = [passage.get("title", None) for passage in hard_negative_context]
+                    hard_negative_ctx_texts = [passage["text"] for passage in hard_negative_context]
+
+                    # all context passages and labels: 1 for positive context and 0 for hard-negative context
+                    ctx_label = [1] * self.num_positives + [0] * self.num_hard_negatives
+                    # featurize context passages
+                    if self.embed_title:
+                        # concatenate title with positive context passages + negative context passages
+                        all_ctx = self._combine_title_context(positive_ctx_titles, positive_ctx_texts) + \
+                                  self._combine_title_context(hard_negative_ctx_titles, hard_negative_ctx_texts)
+                    else:
+                        all_ctx = positive_ctx_texts + hard_negative_ctx_texts
+
+                    # assign empty string tuples if hard_negative passages less than num_hard_negatives
+                    all_ctx += [('', '')] * ((self.num_positives + self.num_hard_negatives) - len(all_ctx))
+
+                    ctx_inputs = self.passage_tokenizer.batch_encode_plus(
+                        all_ctx,
+                        add_special_tokens=True,
+                        truncation=True,
+                        padding="max_length",
+                        max_length=self.max_seq_len_passage,
+                        return_token_type_ids=True
+                    )
 
-            if len(tokenized_passage) == 0:
-                logger.warning(f"The context could not be tokenized, likely because it contains a character that the context tokenizer does not recognize")
-                return None
+                    ctx_segment_ids = np.zeros_like(ctx_inputs["token_type_ids"], dtype=np.int32)
 
-            clear_text["passages"] = positive_context + hard_negative_context
-            tokenized["passages_tokens"] = tokenized_passage
-            features["passage_input_ids"] = ctx_input_ids
-            features["passage_segment_ids"] = ctx_segment_ids
-            features["passage_attention_mask"] = ctx_padding_mask
-            features["label_ids"] = ctx_label
+                    # get tokens in string format
+                    tokenized_passage = [self.passage_tokenizer.convert_ids_to_tokens(ctx) for ctx in ctx_inputs["input_ids"]]
 
+                    if len(tokenized_passage) == 0:
```

Comment:
I think here we would need to iterate over tokenized_passage and check each passage, as tokenized_passage is a list of lists.

Reply:
tokenized strings are only for debugging purposes. So I removed the check alltogether